Sources/StreamChat/Workers/Background/TokenHandler.swift

```diff
@@ -0,0 +1,225 @@
+    func cancelTokenWaiters(with error: Error)
+}
+
+final class DefaultTokenHandler: TokenHandler {
+    private let maximumTokenRefreshAttempts: Int
+    private var retryStrategy: RetryStrategy
+    private let retryTimerType: Timer.Type
+    private var retryTimer: TimerControl?
+    
+    @Atomic private var isRefreshingToken: Bool = false
+    @Atomic private var tokenWaiters: [WaiterToken: TokenWaiter] = [:]
+    
+    private(set) var currentToken: Token?
+    
+    var connectionProvider: UserConnectionProvider
+    
+    // MARK: - Init & Deinit
+    
+    init(
+        connectionProvider: UserConnectionProvider,
+        retryStrategy: RetryStrategy,
+        maximumTokenRefreshAttempts: Int,
+        timerType: Timer.Type
+    ) {
+        self.connectionProvider = connectionProvider
+        self.retryStrategy = retryStrategy
+        self.maximumTokenRefreshAttempts = maximumTokenRefreshAttempts
+        self.retryTimerType = timerType
+    }
+    
+    deinit {
+        let error = ClientError.ClientHasBeenDeallocated()
+        cancelTokenWaiters(with: error)
+    }
+    
+    // MARK: - TokenHandler
+    
+    func set(token: Token, completion: ((Error?) -> Void)?) {
+        if let userId = connectionProvider.userId, token.userId != userId {
+            completion?(ClientError.InvalidToken("The token is for another user"))
+            return
+        }
+        
+        handleTokenResult(.success(token))
+        completion?(nil)
+    }
+    
+    func cancelTokenWaiters(with error: Error) {
+        handleTokenResult(.failure(error))
+    }
+    
+    func refreshToken(completion: @escaping TokenWaiter) {
+        log.debug("Token refresh is initiated.", subsystems: .tokenRefresh)
+        
+        let shouldTriggerRefresh = initiateRefreshIfNotRunning()
+        
+        _ = add(tokenWaiter: completion)
+        
+        guard shouldTriggerRefresh else {
+            log.debug("Token refresh process is already in progress.", subsystems: .tokenRefresh)
+            return
+        }
+        
+        let tokenBeforeRefresh = currentToken
+        
+        retry { [weak self] in
+            guard let self = self else { return }
+            
+            // When `set(token:)` get's invoked while the ongoing token refresh process,
+            // we should assign the given token and cancel the refresh flow. However, cancelling refresh flow
+            // is currently not possible because `tokenProvider` is not returning the cancellable (planned for v5).
+            //
+            // To avoid overriding the token assinged by `set(token:)`, we store the token before the refresh
+            // and compare it we with the current token when refresh is completed. If the tokens does not match,
+            // it means the token was manually assigned and results from `tokenProvider` should be discarded.
+            guard self.currentToken == tokenBeforeRefresh else {
```

Comment:
To correctly handle this, the `tokenProvider` should return the cancellable request token. 

I've added the [note](https://www.notion.so/stream-wiki/StreamChat-v5-e7cd0c287fba41e8af69ae507c40d9c9#2aa41ad748874809926ce39fed3be743) to v5 document so we don't forget